### PR TITLE
DEV-2766 Separate Lilac BAM slicing from calling

### PR DESCRIPTION
**Pull request by nedleitch**

### CODE REVIEW

File: cluster/src/main/java/com/hartwig/pipeline/reruns/StartingPoint.java [hunk 1]
```diff
@@ -66,21 +67,12 @@ enum StartingPoints {
                         LinxSomatic.NAMESPACE,
                         VirusAnalysis.NAMESPACE,
                         Chord.NAMESPACE,
+                        LilacBamSlicer.NAMESPACE,
```
Comment (pauldwolfe) [inline cluster/src/main/java/com/hartwig/pipeline/reruns/StartingPoint.java:70]: I think we want to pull this up to an earlier a bit such that we can separate the bam slicing and lilac calling in reruns.

File: cluster/src/main/java/com/hartwig/pipeline/tertiary/lilac/Lilac.java [hunk 2]
```diff
@@ -29,27 +27,32 @@
...
-    public Lilac(final AlignmentPair alignmentPair, final ResourceFiles resourceFiles, final PurpleOutput purpleOutput) {
-        super(alignmentPair);
+    public Lilac(final LilacBamSliceOutput slicedOutput, final ResourceFiles resourceFiles, final PurpleOutput purpleOutput) {
         this.resourceFiles = resourceFiles;
         PurpleOutputLocations purpleOutputLocations = purpleOutput.outputLocations();
         this.purpleGeneCopyNumber = initialiseOptionalLocation(purpleOutputLocations.geneCopyNumber());
         this.purpleSomaticVariants = initialiseOptionalLocation(purpleOutputLocations.somaticVariants());
+        this.slicedOutput = slicedOutput;
+        this.referenceBam = initialiseOptionalLocation(slicedOutput.reference());
+        this.tumorBam = initialiseOptionalLocation(slicedOutput.tumor());
```
Comment (pauldwolfe) [inline cluster/src/main/java/com/hartwig/pipeline/tertiary/lilac/Lilac.java:55]: do we still need both the sliced and non-sliced bams? I think part of the idea here was to make this stage not depend on anything but the slice bam

File: cluster/src/main/java/com/hartwig/pipeline/tertiary/lilac/LilacBamSlicer.java [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        return Stream.concat(buildCommands(getReferenceBamDownload().getLocalTargetPath(),
+                slicedBam(metadata.reference().sampleName())).stream(),
+                buildCommands(getTumorBamDownload().getLocalTargetPath(), slicedBam(metadata.tumor().sampleName())).stream())
+                .collect(toList());
+
+    }
+
+    @Override
+    public List<BashCommand> referenceOnlyCommands(final SomaticRunMetadata metadata) {
+        return buildCommands(getReferenceBamDownload().getLocalTargetPath(), metadata.sampleName());
```
Comment (pauldwolfe) [inline cluster/src/main/java/com/hartwig/pipeline/tertiary/lilac/LilacBamSlicer.java:60]: why don't we slice for reference and tumor only?

File: cluster/src/main/java/com/hartwig/pipeline/tertiary/lilac/LilacBamSlicer.java [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+        return VmDirectories.outputFile(sampleName + ".hla.bam");
+    }
+
+    private String bai(final String bam) {
+        return bam + ".bai";
+    }
+
+    private AddDatatype[] datatypes(final SomaticRunMetadata metadata) {
+        ArrayList<AddDatatype> datatypes = new ArrayList<>();
+        if (metadata.isSingleSample()) {
```
Comment (pauldwolfe) [inline cluster/src/main/java/com/hartwig/pipeline/tertiary/lilac/LilacBamSlicer.java:120]: Maybe better to use the maybeReferenceSample and maybeTumorSample to decide whether to include the datatype.

**Review by pauldwolfe — CHANGES_REQUESTED** [review]

File: cluster/src/main/java/com/hartwig/pipeline/tertiary/lilac/Lilac.java (hunk 2 above)
Comment (nedleitch) [reply cluster/src/main/java/com/hartwig/pipeline/tertiary/lilac/Lilac.java:55]: Yeah this is just a naming thing, I'm caching the paths to the sliced BAMs as `referenceBam` and `tumorBam`, and keeping a reference to the full `slicedOutput` to access the indices (which I only need in one place). Could just make a member variable for the indices too but felt like it was polluting the namespace a bit.

**Review by pauldwolfe — APPROVED** [review]

**Pull request merged**